text/0000-compose.md

```diff
@@ -0,0 +1,100 @@
+but they all require significant boilerplate and often unreasonable overheard. For example, you may want to split your graphql schemas and resolvers up into many
+separate modular schemas, and combine them onto a single token using schema stiching. However as the number of separate schemas expands upwards of 10, 20, it 
+becomes restrictive to manage separate tokens for each separate schema. For example:
+
+```js
+app.register(TokenA, SchemaA);
+app.register(TokenB, SchemaB);
+app.register(TokenC, SchemaC);
+app.register(TokenD, SchemaD);
+app.register(TokenE, SchemaE);
+
+app.register(SchemaToken, createPlugin({
+  deps: {
+    a: TokenA,
+    b: TokenB,
+    c: TokenC,
+    d: TokenD,
+    e: TokenE,
+  },
+  provides({a, b, c, d, e}) {
+    return mergeSchemas([a,b,c,d,e])
+  }
+}))
+```
+
+This becomes exceedingly difficult as the number of schemas increases. With the app.compose method, this could be replaced with:
+
+```js
+const SchemaToken = createToken('SchemaToken', mergeSchemas);
+app.compose(SchemaToken, SchemaA);
+app.compose(SchemaToken, SchemaB);
+app.compose(SchemaToken, SchemaC);
+app.compose(SchemaToken, SchemaD);
+app.compose(SchemaToken, SchemaE);
+```
+
+# Detailed design
+
+This RFC proposes two nonbreaking API changes which would support managing plugin composition.
+
+1. `createToken` will take a second optional paramter: composeFn which will be used in the case where plugins are registered via `app.compose`.
+2. `app.compose` will be added to the app class which will have the same type signature as `app.register`.
+
+app.compose will throw if the provided token does not have a compose function. During the dependency resolution process, all plugins registered via app.compose
+will be resolved by calling the corresponding compose function. 
+
+app.register overrides any previous values registered on the token, same as its current behavior. 
+
+app.compose overrides a previously registered value via app.register.
+
+Enhancers apply to composed registrations after composition.
+
+## Use cases
+
+- `app.register(TokenA, PluginA)` => TokenA is implemented by PluginA. It may be enhanced in the future, or overwritten via the DI system.
+- `app.compose(TokenA, PluginA)` => PluginA will be composed with other Plugins registered onto TokenA. They may be enhanced in the future, or overwritten via the DI system
+- `app.enhance(TokenA, PluginA)` => PluginA will enhance the existing value on TokenA. This can be commonly used for things like adding caching wrappers. The enhancer is not an implementation of the token.
+
+# Drawbacks
+
+Increased API surface area is the biggest drawback IMO. The addition of this API does create situations where there are multiple possible
+ways of setting up various DI graphs. 
+
+# Alternatives
+
+One alternative is continuing with the current model of managing plugin composition by putting each item onto unique tokens.
```

Comment:
Another alternative is to provide a utility package that exports functionality that does this for you.  It avoids expanding the surface area of `FusionApp` while still cutting back on necessary user-land boilerplate.

Reply:
Potential downside: error messaging and stack traces may be more difficult to hunt down.  Auto-generated tokens aren't great :/

Reply:
Yea, I want to stay away from libraries doing automatic registration and plugin creation and such.

Reply:
For what it's worth, a utility library would let us prove out the value add here with zero pollution of the core API.  And if it doesn't work out, sunsetting a utility library is far less painful.

Reply:
Turns out we are probably going to go with this kind of pattern!